File: python/main/topjslibs/page_js_sources.py
import re

SCRIPT_SRC = re.compile("""<script[^>]+?src=["']([^"']+/)?([^"']+?)["']""")


class PageJsSources:

    def __init__(self, fetch_html=lambda: ""):
        self.fetch_html = fetch_html

    def get(self) -> set[str]:
        html = self.fetch_html()
        match_results = SCRIPT_SRC.findall(html)

        return set([regex_groups[1] for regex_groups in match_results if
                    len(regex_groups) > 1])

File: nodejs/src/topjslibs/PageJsSources.ts
export class PageJsSources {

  private static readonly SCRIPT_SRC =
      new RegExp('<script[^>]+?src=["\'](?:[^"\']+/)?([^"\']+?)["\']', "g");

  private readonly fetchHtml: () => Promise<string>;

  constructor(readonly fetchHtmlLambda: () => Promise<string>) {
    this.fetchHtml = fetchHtmlLambda;
  }

  public async get(): Promise<Set<string>> {
    const html = await this.fetchHtml();
    const matchResults = html.matchAll(PageJsSources.SCRIPT_SRC);
    const result = new Set<string>();

    for (const matchResult of matchResults) {
      if (matchResult.length < 2) {
        continue;
      }

      result.add(matchResult[1]);
    }

    return result;
  }
}
